Add move_toward_angle and angle_difference methods

ettiSurreal opened the pull request:
Closes https://github.com/godotengine/godot-proposals/issues/2074

Adds:

- `angle_difference`, which finds a value between 0 and TAU, finding the shortest path.
- `move_toward_angle`, moves towards an angle by an increment, wrapping around TAU and also finding the shortest path, similar to `lerp_angle`.
- modifies `lerp_angle` code to avoid repetition.

Implementation based on https://github.com/godotengine/godot-proposals/issues/5596#issuecomment-1282661563 and https://github.com/godotengine/godot/pull/37078.

Tested on a custom build, everything seems to be working as it should.

This is my first time making a PR, so let me know what to do if i messed something up!!

marcinn committed:
Implement loading DDS textures at run-time

ettiSurreal commented:
Since 4.0 is in feature freeze and since .NET 6 isn't mentioned in the compiling docs i assume building the C# version isn't fully documented yet, i'll most likely just wait for 4.0 to go in stable before updating the PR again.

adamscoble commented:
@ettiSurreal Oh, this is great to see! Though to stay in line with other functions, should `angle_difference` just be called `angle`? 

Also one other consideration is that in the `move_toward_angle` method, I offset the start value instead of the end, so that the end value is actually reached. Most people will just be testing for the angle anyway, but I do like that idea that the target value is the result of the method, rather than a possible multiple of TAU from it?

Here's the [plugin](https://godotengine.org/asset-library/asset/1549) I made that implements it in case you're curious!

Mickeon committed:
Overhaul Rect2 & Rect2i Documentation

konczg committed:
Add GDExtension support for OpenXR extension wrappers

This commit adds the classes OpenXRExtensionWrapperExtension and OpenXRAPIExtension
that can be used in GDExtensions to define OpenXR extension wrappers.
It modifies extension wrapper registration so that they can be registered
before OpenXRAPI instantiation (e.g. in core level initialization of GDExtensions).

Developed by Migeran (https://migeran.com)

Sauermann committed:
Refactor mouse_entered and mouse_exited notifications

The previous implementation for signals mouse_entered and mouse_exited
had shortcomings that relate to focused windows and pressed mouse buttons.
For example a Control can be hovered by mouse, even if it is occluded by
an embedded window.

This patch changes the behavior, so that Control and Viewport send
their mouse-enter/exit-notifications based solely on mouse position,
visible area, and input restrictions and not on which window has
focus or which mouse buttons are pressed. This implicitly also
changes when the mouse_entered and mouse_exited signals are sent.

This functionality can not be implemented as a part of
Viewport::_gui_input_event, because of its interplay with Windows and
because Viewport::_gui_input_event is based on input and not on
visibility.

KoBeWi committed:
Improve editing of box collision shapes

PrecisionRender commented:
To make things easy to track, make sure to [squash your commits](https://docs.godotengine.org/en/stable/community/contributing/pr_workflow.html#the-interactive-rebase) into a single commit. 👍

Zireael07 commented:
Bump. What's the state of this? Can this go into 4.0 or is this something that's for 4.1?

Calinou commented:
> Bump. What's the state of this? Can this go into 4.0 or is this something that's for 4.1?

Any non-critical change is for 4.1 at the earliest now.

Bestest-Coder committed:
Added --gpu-index to forwardable_cli_arguments

when the gpu index is specified through the CLI, that setting will
be inherited by both the editor (if started through project manager)
and instances of the game started through the editor

Co-authored-by: Rémi Verschelde <rverschelde@gmail.com>

reduz committed:
Support threads in the script debugger

* This implementation adds threads on the side of the client (script debugger).
* Some functions of the debugger are optimized.
* The profile is also now thread safe using atomics.
* The editor can switch between multiple threads when debugging.

This PR adds threaded support for the script language debugger. Every thread has its own thread local data and it will connect to the debugger using multiple thread IDs.
This means that, now, the editor can receive multiple threads entering debug mode at the same time.

AThousandShips commented:
> I hope that's not big of an issue, sorry.

You will have to squash before this can be merged once approved, hopefully you can get it to work

AThousandShips commented:
It's hard, so no worries, I've struggled at the start too!

kleonc commented:
I'll try to rebase. :upside_down_face:

Generally the PR looks ok, I'd say the docs need some more clarification. But I'm not fully sure what's the desired behavior in the first place. E.g. for the current implementation:

```gdscript
move_toward_angle(deg_to_rad(3600), deg_to_rad(90), deg_to_rad(45)) # Returns: deg_to_rad(3645)
move_toward_angle(deg_to_rad(90), deg_to_rad(3600), deg_to_rad(45)) # Returns: deg_to_rad(45)

move_toward_angle(deg_to_rad(0), deg_to_rad(90), deg_to_rad(3600)) # Returns: deg_to_rad(90)
move_toward_angle(deg_to_rad(0), deg_to_rad(90), deg_to_rad(-3600)) # Returns: deg_to_rad(-3600)
```

Is this expected/desired? Or do we maybe want to somehow limit the results to some range (like `[0, TAU)`)? :thinking:

anvilfolk committed:
Add error message when a GDScript resource fails to load.

Currently, GDScripts who are only loaded through `ResourceLoader::load()`,
like Autoloads, do not have a pathway to announce there is an error in their
code. This contributes to significant confusion in error projects when
autoloads are involved. At least partially closes #78230.

ettiSurreal commented:
> I'll try to rebase. 🙃 _Edit: Rebased. I think I haven't changed anything besides the commit name._
> 
Thanks for the help!
> Generally the PR looks ok, I'd say the docs need some more clarification. But I'm not fully sure what's the desired behavior in the first place. E.g. for the current implementation:
> 
> ```gdscript
> move_toward_angle(deg_to_rad(3600), deg_to_rad(90), deg_to_rad(45)) # Returns: deg_to_rad(3645)
> move_toward_angle(deg_to_rad(90), deg_to_rad(3600), deg_to_rad(45)) # Returns: deg_to_rad(45)
> 
> move_toward_angle(deg_to_rad(0), deg_to_rad(90), deg_to_rad(3600)) # Returns: deg_to_rad(90)
> move_toward_angle(deg_to_rad(0), deg_to_rad(90), deg_to_rad(-3600)) # Returns: deg_to_rad(-3600)
> ```
> 
> Is this expected/desired? Or do we maybe want to somehow limit the results to some range (like `[0, TAU)`)? 🤔

As for the docs i am not the best at writing those, I think someone might rewrite this sometime.
As for the behavior, it's the same logic as `lerp_angle`, I'll do some tests tomorrow to see if i messed something up.

AThousandShips commented on modules/mono/glue/GodotSharp/GodotSharp/Core/Mathf.cs:
```diff
@@ -1197,6 +1219,38 @@ public static double MoveToward(double from, double to, double delta)
+        public static float MoveTowardAngle(float from, float to, float delta)
+        {
+            float diff = AngleDifference(from, to);
+            return from + Min(Abs(diff), delta) * Sign(diff);
```
It looks like other methods use `Math.Abs` and so on, is that the standard approach here?

ettiSurreal replied:
Fixed!

adamscoble commented:
> As for the docs i am not the best at writing those, I think someone might rewrite this sometime. As for the behavior, it's the same logic as `lerp_angle`, I'll do some tests tomorrow to see if i messed something up.

In case it helps, I wrote my function docs (and names) to mirror Godot's, so feel free to pinch [them](https://github.com/adamscoble/godot-math-extensions/blob/main/addons/math_extensions/math_extensions.gd) if you'd like!

Also regarding @kleonc's comment, in my GDScript implementation the result of `move_toward_angle` is always `to`, which I believe is more intuitive than `to + TAU * x`.

PrecisionRender commented:
@adamscoble I just straight-up copied the `lerp_angle` function code in my implementation.

````
static func move_toward_angle(from: float, to: float, delta: float) -> float:
	var difference = fmod(to - from, TAU)
	var distance = fmod(2.0 * difference, TAU) - difference
	return move_toward(from, from + distance, delta)
````

adamscoble commented:
> @adamscoble I just straight-up copied the `lerp_angle` function code in my implementation.
Oh, well there you go! I don't use the lerp functions often for this stuff, as I created a [float smoothing class](https://github.com/adamscoble/godot-math-extensions/blob/main/addons/math_extensions/float_smooth.gd) in that same repository to avoid the framerate-dependent smoothing of lerp. So it seems you're being consistent with Godot!

kleonc commented:
> Also regarding @kleonc's comment, in my GDScript implementation the result of `move_toward_angle` is always `to`, which I believe is more intuitive than `to + TAU * x`.

@adamscoble Your implementation behaves the same as the one in this PR, note the examples I gave in https://github.com/godotengine/godot/pull/69081#issuecomment-1601200040 are kinda edge cases (like why would you want to move away by 3600 degrees?). For both implementations if `delta < 0` the result is not being clamped whatsoever.

For the the `move_toward`/`Vector2.move_toward`/`Vector3.move_toward` it makes that `delta < 0` results in moving away from `to` without any clamping/limiting. The relevant 1D/2D/3D spaces are conceptually unlimited, meaning you can always move further away.
But "angle space" is different, it's periodic. You can't really move further away than 180 degrees from the given angle. So maybe `move_toward_angle` for `delta < 0` should be clamped/limited to `[to - PI, to + PI]` range (whether it should be wrapped to a range like `[0, TAU)` or `[-PI, PI)` is a different question).

As a simple example, what should be the result of:
```gdscript
move_toward_angle(deg_to_rad(90), deg_to_rad(0), deg_to_rad(-180))
```
Which can be interpreted as: we're at 90 degrees and we're moving 180 degrees away from 0 degrees.
Currently the result would be 270 degrees. I say maybe it should be clamped so the result would be 180 degrees (because when moving past 180 degrees we would start getting closer to the 0 degrees we were meant to be moving away from).

But maybe the current behavior is desired, I don't know. What I'm saying is that firstly the expected behavior should be well defined for all possible inputs. Tweaking the implementation/docs accordingly is the next step.

KoBeWi committed:
Add separate editor plugin for TileMap and TileSet

PrecisionRender commented:
> But maybe the current behavior is desired, I don't know. What I'm saying is that firstly the expected behavior should be well defined for all possible inputs. Tweaking the implementation/docs accordingly is the next step.

I'd say it should work the same as `lerp_angle`, for the sake of consistency.

ettiSurreal commented:
Tested some things out in a GDScript port of the functions.
Clamping the value between [0.0, TAU]/[-PI, PI] seems to be as simple as using a wrap function on the return, I guess i should add that?
Negative values are seem to be an issue. Currently (if the wrap function is added),
> ```gdscript
> move_toward_angle(deg_to_rad(90), deg_to_rad(0), deg_to_rad(-180))
> ```
would return -90 degrees, since negative values currently do not support stopping at at the opposite angle. Which would mean if you run it per frame it will oscillate.
However, if you wanna move toward to the opposite angle you can just add 180 degrees to p_to and use a positive delta. I found a simple way to make it so the function does it automatically when you input a negative value, but it requires an If statement at the start of the function. Unity's equivalent function omits automatically doing it for optimization reasons, should we as well?

raulsntos commented on modules/mono/glue/GodotSharp/GodotSharp/Core/Mathf.cs:
```diff
@@ -103,6 +103,32 @@ public static double Acos(double s)
             return Math.Acos(s);
         }
 
+        /// <summary>
+        /// Gets the difference between two angles (in radians) by a normalized value,
+        /// but wraps around <see cref="Tau"/>, also creating a shortest path.
+        /// </summary>
```
The C# documentation is different from the GDScript documentation. We try to keep them in sync.

ettiSurreal replied:
Wrote a new doc entirely and copied it over to C#.

raulsntos commented on modules/mono/glue/GodotSharp/GodotSharp/Core/Mathf.cs:
```diff
@@ -103,6 +103,32 @@ public static double Acos(double s)
+        /// but wraps around <see cref="Tau"/>, also creating a shortest path.
+        /// </summary>
+        /// <param name ="from">The start angle.</param>
+        /// <param name ="to">the destination angle.</param>
```
```suggestion
        /// <param name ="to">The destination angle.</param>
```

raulsntos commented on modules/mono/glue/GodotSharp/GodotSharp/Core/Mathf.cs:
```diff
@@ -103,6 +103,32 @@ public static double Acos(double s)
+        /// but wraps around <see cref="Tau"/>, also creating a shortest path.
+        /// </summary>
+        /// <param name ="from">The start angle.</param>
+        /// <param name ="to">the destination angle.</param>
```
```suggestion
        /// <param name ="to">The destination angle.</param>
```

raulsntos commented on modules/mono/glue/GodotSharp/GodotSharp/Core/Mathf.cs:
```diff
@@ -1197,6 +1219,38 @@ public static double MoveToward(double from, double to, double delta)
+        /// Similar to <see cref="MoveToward"/> but wraps around <see cref="Tau"/> and creates a shortest path.
+        /// </summary>
+        /// <param name="from">The start Angle.</param>
+        /// <param name="to">The Angle to move towards.</param>
```
```suggestion
        /// <param name="to">The angle to move towards.</param>
```

raulsntos commented on modules/mono/glue/GodotSharp/GodotSharp/Core/Mathf.cs:
```diff
@@ -1197,6 +1219,38 @@ public static double MoveToward(double from, double to, double delta)
+        /// <param name="from">The start Angle.</param>
+        /// <param name="to">The Angle to move towards.</param>
+        /// <param name="delta">The amount to move by.</param>
+        /// <returns>The Angle after moving.</returns>
```
```suggestion
        /// <returns>The angle after moving.</returns>
```

raulsntos commented on modules/mono/glue/GodotSharp/GodotSharp/Core/Mathf.cs:
```diff
@@ -1197,6 +1219,38 @@ public static double MoveToward(double from, double to, double delta)
+        /// Similar to <see cref="MoveToward"/> but wraps around <see cref="Tau"/> and creates a shortest path.
+        /// </summary>
+        /// <param name="from">The start Angle.</param>
+        /// <param name="to">The Angle to move towards.</param>
```
```suggestion
        /// <param name="to">The angle to move towards.</param>
```

raulsntos commented on modules/mono/glue/GodotSharp/GodotSharp/Core/Mathf.cs:
```diff
@@ -1197,6 +1219,38 @@ public static double MoveToward(double from, double to, double delta)
+        /// <param name="from">The start Angle.</param>
+        /// <param name="to">The Angle to move towards.</param>
+        /// <param name="delta">The amount to move by.</param>
+        /// <returns>The Angle after moving.</returns>
```
```suggestion
        /// <returns>The angle after moving.</returns>
```

ettiSurreal commented on modules/mono/glue/GodotSharp/GodotSharp/Core/Mathf.cs:
```diff
@@ -103,6 +103,32 @@ public static double Acos(double s)
+        public static double AngleDifference(double from, double to)
+        {
+            double difference = (to - from) % Mathf.Tau;
+            return ((2 * difference) % Mathf.Tau) - difference;
```
Fixed!

ettiSurreal commented:
Wrapped the angle between [-PI, PI], since that's how `lerp_angle` seems to work.
For now decided to keep negative delta as "unsupported" for optimization (let me know if i should do otherwise), clarifying it in the docs, which I also just rewrote.

AThousandShips commented on core/math/math_funcs.h:
```diff
@@ -427,13 +432,23 @@ class Math {
+	}
+	static _ALWAYS_INLINE_ float move_toward_angle(float p_from, float p_to, float p_delta) {
+		float difference = angle_difference(p_from, p_to);
+		return wrapf(p_from + fmin(abs(difference), p_delta) * SIGN(difference), -(float)Math_PI, (float)Math_PI);
```
```suggestion
		return Math::wrapf(p_from + MIN(Math::abs(difference), p_delta) * SIGN(difference), -(float)Math_PI, (float)Math_PI);
```
We're explicit about the methods from this same class, also use the Godot specific minimum function not the global scope one

Same applies for the line above

AThousandShips replied:
I'd also suggest prefixing our own functions here with `Math::` as that's partially adhered to

ettiSurreal replied:
Done, I've seen a little bit of inconsistency with the former but that's probably for another PR.

kleonc commented:
>Wrapped the angle between [-PI, PI], since that's how `lerp_angle` seems to work.

@ettiSurreal `lerp_angle` doesn't do any wrapping (and its behavior should remain unchanged as it would be a compatibility breaking change). Hence I think leaving `move_toward_angle` without wrapping the result to a specific range is the way to go. If desired, user can always wrap it manually, e.g. `wrapf(move_toward_angle(from, to, delta), -PI, PI)`.

>For now decided to keep negative delta as "unsupported" for optimization (let me know if i should do otherwise), clarifying it in the docs, which I also just rewrote.

Performance shouldn't differ much, I think the implementation would be as simple as:
```cpp
static _ALWAYS_INLINE_ double move_toward_angle(double p_from, double p_to, double p_delta) {
	double difference = angle_difference(p_from, p_to);
	double abs_difference = abs(difference);
	// When `delta < 0` move no further than to PI radians away from `p_to` (as PI is the max possible angle distance).
	return p_from + CLAMP(delta, abs_difference - Math_PI, abs_difference) * SIGN(difference);
}
```

---
 
> I'd say it should work the same as `lerp_angle`, for the sake of consistency.

@PrecisionRender Agree consistency is good but the thing is these already work differently as `lerp_angle` is not limiting the result in any direction. You can extrapolate by passing either `weight > 1` or `weight < -1`, e.g. `lerp_angle(0, PI, 99)` returns `99 * PI`.
To me `move_toward_angle` conceptually behaves differently, it's main purpose is to perform logic limiting the movement against the `to` parameter. Hence when moving away it makes sense to me to not move past 180 degrees away from `to`. So I'd limit it like that. But if others disagree and it's more desired to not limit moving away anyhow (possibly going full circles), it would be fine to me too. :upside_down_face:

ettiSurreal commented:
> > Wrapped the angle between [-PI, PI], since that's how `lerp_angle` seems to work.
> 
> @ettiSurreal `lerp_angle` doesn't do any wrapping (and its behavior should remain unchanged as it would be a compatibility breaking change). Hence I think leaving `move_toward_angle` without wrapping the result to a specific range is the way to go. If desired, user can always wrap it manually, e.g. `wrapf(move_toward_angle(from, to, delta), -PI, PI)`.

I must've mistested it, I thought `lerp_angle` by itself stayed between [-PI, PI] unless you overloaded the weight.

> > For now decided to keep negative delta as "unsupported" for optimization (let me know if i should do otherwise), clarifying it in the docs, which I also just rewrote.
> 
> Performance shouldn't differ much, I think the implementation would be as simple as:
> 
> ```c++
> static _ALWAYS_INLINE_ double move_toward_angle(double p_from, double p_to, double p_delta) {
> 	double difference = angle_difference(p_from, p_to);
> 	double abs_difference = abs(difference);
> 	// When `delta < 0` move no further than to PI radians away from `p_to` (as PI is the max possible angle distance).
> 	return p_from + CLAMP(delta, abs_difference - Math_PI, abs_difference) * SIGN(difference);
> }
> ```
Since that's seemingly not a concern I can add your implementation then. My own implementation would've been something like this.
```c++
static _ALWAYS_INLINE_ double move_toward_angle(double p_from, double p_to, double p_delta) {
	if (p_delta > 0.0) {
		double difference = Math::angle_difference(p_from, p_to);
		return Math::wrapf(p_from + MIN(Math::abs(difference), p_delta) * SIGN(difference), -Math_PI, Math_PI);
	} else {
		double difference = Math::angle_difference(p_from, p_to + Math.PI);
		return Math::wrapf(p_from + MIN(Math::abs(difference), Math::abs(p_delta)) * SIGN(difference), -Math_PI, Math_PI);		
```

PrecisionRender commented:
> @PrecisionRender Agree consistency is good but the thing is these already work differently as `lerp_angle` is not limiting the result in any direction. You can extrapolate by passing either `weight > 1` or `weight < -1`, e.g. `lerp_angle(0, PI, 99)` returns `99 * PI`.
> To me `move_toward_angle` conceptually behaves differently, it's main purpose is to perform logic limiting the movement against the `to` parameter. Hence when moving away it makes sense to me to not move past 180 degrees away from `to`. So I'd limit it like that. But if others disagree and it's more desired to not limit moving away anyhow (possibly going full circles), it would be fine to me too. 🙃

I've never run into a scenario where I need to do that moving _away_ from the desired angle, so I imagine both implementations would work.

dsnopek committed:
Allow resizing String's from GDExtension

smix8 committed:
Add NavigationServer API to enable regions and links

Adds NavigationServer API to enable regions and links.

BastiaanOlij committed:
Expose RenderingSceneBuffers through ClassDB

raulsntos committed:
C#: Print error when MethodBind/Callable call fails

KoBeWi committed:
Assume root when dropping node to unassigned script

BastiaanOlij committed:
Add custom texture create function

akien-mga committed:
CI: Allow skipping our GHA workflows with `DISABLE_GODOT_CI` variable

Useful for custom forks of Godot which don't want to run our CI for any
reason.

This is configured in `settings/variables/actions` for the repository,
setting it to any value aside from an empty string will skip all jobs.

smix8 committed:
Add detail to NavigationAgent signal descriptions

Adds detail to NavigationAgent signal descriptions.

Calinou committed:
Fill in descriptions for import options in the class reference

KoBeWi committed:
Rework modifying tile source ID

AncientSion commented:
can you add this to 3.5/6, i need it badly.

AThousandShips commented:
@AncientSion Please don't bump PRs without contributing significant new information. Use the :+1: reaction button on the first post instead.

AThousandShips commented:
This being a new feature it can be considered for cherry-picking, unsure if that will be done though

timothyqiu commented:
Any reason why the function is named `move_toward_angle` instead of `rotate_toward`? I think it's technically not possible to _move_ toward an angle :P

timothyqiu commented on doc/classes/@GlobalScope.xml:
```diff
@@ -72,6 +72,14 @@
+			<param index="0" name="from" type="float" />
+			<param index="1" name="to" type="float" />
+			<description>
+				Returns the difference between the two angles, correctly wrapping around [constant @GDScript.TAU].
```
```suggestion
				Returns the difference between the two angles. in the range of [code][-PI, +PI][/code]. When [param from] and [param to] are opposite, returns [code]-PI[/code] if [param from] is smaller than [param to], or [code]PI[/code] otherwise.
```

PrecisionRender commented:
> Any reason why the function is named `move_toward_angle` instead of `rotate_toward`? I think it's technically not possible to _move_ toward an angle :P

I don't disagree, but I think these are the current reasons:

a. To be consistent with the `lerp_angle()` function
b. To be consistent with other engines

timothyqiu commented:
> a. To be consistent with the `lerp_angle()` function

Make sense. But to be fair, "lerp" is one term (same goes for "cubic_interpolate"). The only choice for naming a rotation variant is adding a prefix or suffix :stuck_out_tongue: 

> b. To be consistent with other engines

Unity has inconsistent naming: the standalone version is named `MoveTowardsAngle`, while the member version is named `RotateTowards`.

kleonc commented:
> Unity has inconsistent naming: the standalone version is named `MoveTowardsAngle`, while the member version is named `RotateTowards`.

And there's only `Vector3.RotateTowards`, no `Vector2.RotateTowards` or `Vector4.RotateTowards`.

Regarding the behavior for negative delta:
- [`MoveTowardsAngle`](https://docs.unity3d.com/ScriptReference/Mathf.MoveTowardsAngle.html):
>For optimization reasons, negative values of maxDelta are not supported and may cause oscillation. To push current away from a target angle, add 180 to that angle instead.
- [`Vector3.RotateTowards`](https://docs.unity3d.com/ScriptReference/Vector3.RotateTowards.html):
> If a negative value is used for maxRadiansDelta, the vector will rotate away from target/ until it is pointing in exactly the opposite direction, then stops.

So not only the naming is inconsistent. :upside_down_face:

---

Note that adding `Vector3.RotateTowards`-like method to `VectorN` types is also desired/requested, see: https://github.com/godotengine/godot-proposals/discussions/7302.

Commit:
Fix npe when registering the sensors.

reduz committed:
Add ability to call code on rendering thread

As more users use compute in Godot 4, the way they do is most likely incompatible when running
on separate threads and will start erroring soon as we improve the thread safety of the render thread.

To properly run code on the render thread, this function was added. Use like this:

```GDScript

func initialize_compute_code():
	....

func update_compute_code(custom_data):
	...

func _ready():
	RenderingServer.call_on_render_thread( initialize_compute_code )

func _process():
	RenderingServer.call_on_render_thread( update_compute_code.bind(with_data) )

```

sfreed141 committed:
Respect "editor/naming/node_name_casing" in Create New Scene dialog

aaronfranke committed:
Improve overriding the root type or root name in the scene importer

aaronfranke committed:
Cosmetic changes in GLTF node generation code

Jamsers commented:
For anyone stumbling in here looking for an easy to use rotate_toward for both quats and basis' in the meantime, I used these:
```
func rotate_toward(from: Quaternion, to: Quaternion, delta: float) -> Quaternion:
	return from.slerp(to, min(1.0, delta / from.angle_to(to)))

func basis_rotate_toward(from: Basis, to: Basis, delta: float) -> Basis:
	return Basis(rotate_toward(from.get_rotation_quaternion(), to.get_rotation_quaternion(), delta))
```
Thanks to [kleonc](https://www.reddit.com/r/godot/comments/156gmx8/comment/jszze4p/) from Reddit for the quick fix ❤️

YeldhamDev committed:
Fix some issues with scene tabs

- Stop scene creation on double-click when hovering the offset buttons.
- Fix the "Add Scene" button's position on RTL mode.

adamscott committed:
Fix file permissions for the web platform

LRFLEW committed:
Fix transparent viewport backgrounds with custom clear color

dalexeev committed:
GDScript: Fix bug with identifier shadowed below in current scope

KoBeWi committed:
Improve and clarify paused Tweens

clayjohn committed:
Check if shader cache directory is available before using cache

yvie-k committed:
Fix dangling pointers in _clean_up_swap_chain

starry-abyss committed:
Project manager button rearrangement

YuriSizov committed:
Merge pull request #75620 from jpcerrone/fix_shadow_warnings_not_going_away_after_ignoring

Fix for not being able to ignore shadowing warnings on class scope

YuriSizov committed:
Merge pull request #78365 from smix8/ps_navmap_default_up_4.x

Add ProjectSettings navigation map default up

YuriSizov committed:
Merge pull request #79257 from DeeJayLSP/sharpp

Lossy WebP: Enable sharp RGB to YUV conversion

YuriSizov committed:
Merge pull request #79492 from dsnopek/gdextension-remove-editor-plugins-at-shutdown

Prevent GDExtensions from trying to remove editor plugins at shutdown

YuriSizov committed:
Merge pull request #79603 from bitsawer/fix_instance_uniform_update

Fix instance uniform data buffer update delay

YuriSizov committed:
Merge pull request #79720 from clayjohn/Particles-emitting-docs

Add detail to emitting docs for particles

YuriSizov committed:
Merge pull request #79734 from somecrepe/master

Change the text for the flat button preview to follow pattern

YuriSizov committed:
Merge pull request #79764 from kulkalkul/fix_docs_set_multiplayer_authority_propagate

Clarify `set_multiplayer_authority` documentation regarding propagation

YuriSizov committed:
Merge pull request #79786 from smix8/navmesh_max_edge_len_4.x

Disable NavigationMesh `edge_max_length` property by default

YuriSizov committed:
Merge pull request #79794 from DarthKitten2130/master

Update outdated C# code sample in `AStarGrid2D` documentation

YuriSizov committed:
Merge pull request #79812 from ParsleighScumble/parsleigh/init-sky-scene-state

Use defaults to initialize sky data in case of no sky

YuriSizov committed:
Merge pull request #79865 from clayjohn/discard-shader

Fix use of discard in shaders

YuriSizov committed:
Merge pull request #79866 from adamscott/fix-file-permissions-web

Fix file permissions for the web platform (affects every Unix-like platform)

YuriSizov committed:
Merge pull request #79239 from raulsntos/dotnet/document-generated-members

C#: Document generated members

YuriSizov committed:
Merge pull request #79736 from timothyqiu/bit

Fix typo in ResourceImporterImageFont

YuriSizov committed:
Merge pull request #79748 from Repiteo/dotnet-documentation-fixes

Fix issues in C# documentation comments

YuriSizov committed:
Merge pull request #79793 from trollodel/hide_make_floating_swm

Make the single window mode check more strict

YuriSizov committed:
Merge pull request #79879 from KoBeWi/the_tweenld

Improve and clarify paused Tweens

YuriSizov committed:
Merge pull request #79884 from Dragoncraft89/master

Vulkan: Fix dangling pointers in `_clean_up_swap_chain`

paddy-exe committed:
Additional Visual Shader Nodes

* RotationByAxis Visual Shader Node added
* WorldPositionFromDepth Visual Shader Node added
* ScreenNormalWorldSpace Visual Shader Node added

KoBeWi committed:
Improve atlas tile size dragging

KoBeWi committed:
Add help label about creating multiple/big tiles

YeldhamDev committed:
Fix spacing between icon and "Output" button

kdiduk committed:
[bugfix] Fix reimporting scene withh default values selected

This commit fixes #78140

When the scene was re-imported with non-default values of some settings, re-importing it again using default values for those settings didn't have the effect.

The problem was that when handling the reimport, a wrong dictionary of the settings was used.

BastiaanOlij committed:
Add exceptions for breakage introduced in RD barriers and prevent future breakage

bruvzg committed:
[Label] Remove extra line spacing from Label minimum size calculations.

smix8 committed:
Update NavObstacle creation to new Mutex style

Updates NavObstacle creation to new Mutex style.

dalexeev committed:
GDScript: Add validation for `@export_node_path` annotation arguments

Co-authored-by: George Marques <george@gmarqu.es>

timothyqiu committed:
Sort system font menu in Inspector

YuriSizov committed:
Merge pull request #54524 from Calinou/animation-editor-add-bypass-confirmation

Skip keyframe creation dialog when holding Shift in the animation editor

YuriSizov committed:
Merge pull request #72346 from myaaaaaaaaa/disconnect-order

Avoid sorting CallableCustomMethodPointers by their actual address values

YuriSizov committed:
Merge pull request #72664 from paddy-exe/additional-vs-nodes

Add more useful Visual Shader nodes

YuriSizov committed:
Merge pull request #76582 from reduz/threaded-debugger

Support threads in the script debugger

YuriSizov committed:
Merge pull request #78615 from RandomShaper/fix_doc_cache

Re-enable docs cache with fixes

YuriSizov committed:
Merge pull request #78747 from RandomShaper/fix_trans_threading

Support loading of translations on threads

YuriSizov committed:
Merge pull request #78827 from Rindbee/fix-history-mismatch

Fix history mismatch

YuriSizov committed:
Merge pull request #78987 from RandomShaper/err_bad_deferred_target

Let user know about dead instances in deferred calls

YuriSizov committed:
Merge pull request #78988 from RandomShaper/fix_change_scene

Reimplement scene change

YuriSizov committed:
Merge pull request #79129 from smix8/navbase_enabled_4.x

Add NavigationServer API to enable regions and links

YuriSizov committed:
Merge pull request #79203 from KoBeWi/visible_confusion

Don't grab theme icons for scripts

YuriSizov committed:
Merge pull request #79248 from Sauermann/fix-svc-focus-propagation

Make `SubViewportContainer` event propagation aware of focused Control

YuriSizov committed:
Merge pull request #79288 from BastiaanOlij/custom_texture_api

Add custom texture create function

YuriSizov committed:
Merge pull request #79351 from Repiteo/dotnet-handle-warnings

C#: Fix MSVC dotnet builds failing if running `dev_mode`

YuriSizov committed:
Merge pull request #79401 from smix8/navagent_desc_4.x

Add detail to NavigationAgent signal descriptions

YuriSizov committed:
Merge pull request #79443 from Sauermann/fix-ingore-hovering

Fix physics passive hovering with `MOUSE_FILTER_IGNORE`

YuriSizov committed:
Merge pull request #79683 from Zylann/fix_gdext_get_property_list

GDExtension: Fix `_get_property_list` not working correctly in parent classes

YuriSizov committed:
Merge pull request #79763 from mihe/indexed-properties

Add support for indexed properties in GDExtension

YuriSizov committed:
Merge pull request #79766 from Sauermann/fix-window-determinant-zero

Add determinant check for minimized windows

YuriSizov committed:
Merge pull request #79805 from kumikumi/fix-subviewport-cursor

Prevent SubViewportContainer overriding Subviewport's cursor with its own cursor

YuriSizov committed:
Merge pull request #79883 from clayjohn/GLES3-shader-cache-dir

Check if shader cache directory is available before using cache

YuriSizov committed:
Merge pull request #79899 from KoBeWi/snapped_in_half

Improve atlas tile size dragging

YuriSizov committed:
Merge pull request #79913 from bruvzg/lbl_min_h_line

[Label] Remove extra line spacing from Label minimum size calculations.

YuriSizov committed:
Merge pull request #79916 from smix8/navobstacle_mutex_4.x

Update NavObstacle creation to new Mutex style

Sauermann committed:
Add note about mouse movement input events in `MouseFilter` constants

Sauermann committed:
Disable error condition for accessing safe rect

This PR removes the error-message, which is currently expected to happen.
The task of correctly fixing the underlying problem of window popup
closing is more complex, so this PR solves the immediate
error-notification-problem.

YuriSizov committed:
Merge pull request #77537 from henriquelalves/feature/ei_save_all_scenes

Expose `save_all_scenes` method to EditorInterface

KoBeWi committed:
Properly clear scene tiles

hvarga committed:
Fix scene tab close

hvarga committed:
Fix current-scene update

Ymanawat committed:
Fixed Tree Selection Bug

bruvzg committed:
[macOS Export] Disable unpacked .app bundle export on Windows.

RandomShaper committed:
Revert "Implement parallel foreach() for easier multithreading"

This reverts commit e28868e30cd478e878d0a8331e17e2cf90039558.

RandomShaper committed:
Fix life cycle of ResourceImporterTexture not tracked properly

YuriSizov committed:
Merge pull request #63826 from deakcor/dev_clipboard

Add `clipboard_has/get_image` methods to DisplayServer

YuriSizov committed:
Merge pull request #68259 from konczg/openxr_extension_wrapper_gdextension

Add GDExtension support for OpenXR extension wrappers

YuriSizov committed:
Merge pull request #79142 from BastiaanOlij/register_render_buffers

Expose RenderSceneBuffers(RD) through ClassDB

YuriSizov committed:
Merge pull request #79156 from dsnopek/gdextension-string-resize

Allow resizing Strings from GDExtension

YuriSizov committed:
Merge pull request #79512 from KoBeWi/tile_expanded_universe

Add option to expand tile polygon editors

YuriSizov committed:
Merge pull request #79904 from KoBeWi/how_to_create_huge_tiles_using_a_simple_trick

Add help label about creating multiple/big tiles

YuriSizov committed:
Merge pull request #79908 from YeldhamDev/output_icon_space

Fix spacing between icon and "Output" button

YuriSizov committed:
Merge pull request #79928 from timothyqiu/fonts

Sort system font menu in Inspector

YuriSizov committed:
Merge pull request #79934 from Sauermann/fix-mousefilter-doc

Add note about mouse movement input events in `MouseFilter` constants

YuriSizov committed:
Merge pull request #79937 from Sauermann/fix-expected-safe-rect-error

Disable error condition for accessing safe rect

YuriSizov committed:
Merge pull request #78944 from bgie/particle_mask

Add option to center image when loading particle emission mask

YuriSizov committed:
Merge pull request #79953 from RandomShaper/revert_wtp_for_range

Revert "Implement parallel foreach() for easier multithreading"

YuriSizov committed:
Merge pull request #79954 from RandomShaper/fix_res_imp_life

Fix life cycle of ResourceImporterTexture not tracked properly

viksl committed:
Update C# example of tween_method in Tween class with a parameter to the lambda method

YuriSizov committed:
Merge pull request #79941 from KoBeWi/scene_tile_devastation

Properly clear scene tiles

YuriSizov committed:
Merge pull request #74717 from KoBeWi/tilesetmap_handler_plugin

Add separate editor plugin for TileMap and TileSet

YuriSizov committed:
Merge pull request #79405 from Calinou/doc-import-options

Fill in descriptions for import options in the class reference

Owl-A committed:
added state sync after call to _integrate_forces

RandomShaper committed:
Fix life cycle of ResourceImporterTexture better

detomon committed:
Initialize MSDF parameters in BaseMaterial3D with default

KoBeWi committed:
Reverse condition for skipping directories

clayjohn committed:
Merge pull request #79983 from detomon/initialize-msdf-parameters-in-base-material

Initialize MSDF parameters in BaseMaterial3D with default

vnen committed:
GDScript: Optimize operators by assuming the types

This assumes that operators are called usually with the same type of
operands as the first time. So it stores the types of the first run and
if matched it uses an optimized path by calling the validated operator
function directly. Otherwise it uses the regular untyped evaluator.

With this change, if operators do use the same type they run quite
faster. OTOH, if the types mismatch it takes longer to run than they
would with the previous code.

aaronfranke committed:
Merge pull request #79981 from RandomShaper/fix_res_imp_life_2

Fix life cycle of ResourceImporterTexture better

dalexeev committed:
GDScript: Add constant string support for POT generator

darksylinc committed:
Fix uninitialized variable ending up sent to Vulkan

The first time a shader is compiled Godot performs the following:

```cpp
for (uint32_t i = 0; i < SHADER_STAGE_MAX; i++) {
	if
(spirv_data.push_constant_stages_mask.has_flag((ShaderStage)(1 << i))) {
		binary_data.push_constant_vk_stages_mask |=
shader_stage_masks[i];
	}
}
```

However binary_data.push_constant_vk_stages_mask is never initialized to
0 and thus contains garbage data or'ed with the good data.

This value is used by push constants (and many other things) thus it can
be a big deal.

Fortunately because the relevant flags are always guaranteed to be set
(but not guaranteed to be unset), the damage is restricted to:

1. Performance (unnecessary flushing & over-excessive barriers)
2. Overwriting push descriptors already set (this would be serious,
doesn't seem to be an issue)
3. Driver implementations going crazy when they see bits set they don't
expect (unknown if this is an issue)

This uninitialized value is later saved into the binary cache.

Valgrind is able to detect this bug on the first run, but not on the
subsequent ones because they data comes from a file.

cache_file_version has been bumped to force rebuild of all cached
shaders. Because the ones generated so far are compromised.

KoBeWi committed:
Edit TileSet source on double click

AbeniMatteo committed:
Avoid dictionary lookup for common colors

dalexeev committed:
Editor: Fix escaping issues with POT generator

clayjohn committed:
Merge pull request #80034 from darksylinc/matias-fixes

Fix uninitialized variable ending up sent to Vulkan

timothyqiu committed:
Avoid punning [param enabled] in documentation

clayjohn committed:
Enable depth writes during shadow pass and depth pass. Disable during color pass

This fixes an unreported performance bug. Basically, without this change, the depth prepass did nothing

darksylinc committed:
Fix validation layer warnings

1. Validation layers on Windows were complaining w/
VUID-VkSwapchainCreateInfoKHR-surface-01270 that we were not calling
vkGetPhysicalDeviceSurfaceSupportKHR before vkCreateSwapchainKHR.

2. Godot was only calling vkGetPhysicalDeviceSurfaceSupportKHR at
startup, but it should be doing this for every window w/ a new surface
it wants to create, not just the first one.
- In practice this will likely not make a difference. If
vkGetPhysicalDeviceSurfaceSupportKHR returns false after initialization,
there's nothing we can do about it and it is likely because something
else went terribly wrong, which is why the error message is worded like
that.
- This is mostly to shut up validation layers. Though technically,
the layers are right.

3. Do not call vkGetPhysicalDeviceSurfaceSupportKHR on queues we don't
even plan on ever using. We don't know how drivers will react to that
(e.g. they may preemptetively allocate resources to support presentation
on exotic queues, instead of just saying no). Just behave like every
other Vulkan app out there.

clayjohn committed:
Revert "Let user know about dead instances in deferred calls"

This reverts commit 3a6527d6d3c15bc052b9f27ccb1b4ad7c10c135f.

YuriSizov committed:
Merge pull request #80081 from clayjohn/revert-dead-error

Revert "Let user know about dead instances in deferred calls"

AThousandShips committed:
Fix API hash related crash in `EditorSettings`

neikeq committed:
Merge pull request #73257 from RedworkDE/net-android-support

C#: Support exporting for Android

YuriSizov committed:
Merge pull request #50674 from starry-abyss/master

Reorganize buttons in the project manager

YuriSizov committed:
Merge pull request #73226 from AThousandShips/gdscript_tok_improvement

Improve GDScript identifier tokenization

YuriSizov committed:
Merge pull request #78254 from dalexeev/gds-fix-property-group-name-conflict

GDScript: Fix conflict between property and group names

YuriSizov committed:
Merge pull request #78941 from dalexeev/gds-doc-comments-deprecated-and-experimental

GDScript: Add `@deprecated` and `@experimental` doc comment tags

YuriSizov committed:
Merge pull request #79358 from timothyqiu/indentation

Make indentation indicators translatable

YuriSizov committed:
Merge pull request #79610 from aaronfranke/toggle-comment-slash

Add Ctrl+/ as a shortcut to toggle comment in addition to Ctrl+K

YuriSizov committed:
Merge pull request #79696 from reduz/call-on-render-thread

Add ability to call code on rendering thread

YuriSizov committed:
Merge pull request #79681 from m4gr3d/fix_npe_main

Fix NullPointerException when registering the sensors

YuriSizov committed:
Merge pull request #79737 from Calinou/resourceloader-load-error-clarify-expected-type

Mention expected resource type in ResourceLoader load error

YuriSizov committed:
Merge pull request #79876 from LRFLEW/mobalpha

Fix transparent viewport backgrounds with custom clear color

YuriSizov committed:
Merge pull request #79880 from dalexeev/gds-fix-id-shadowing-below

GDScript: Fix bug with identifier shadowed below in current scope

YuriSizov committed:
Merge pull request #79917 from hvarga/fix-scene-tab-close

Fix crash when using "Close All Tabs"

YuriSizov committed:
Merge pull request #79945 from hvarga/fix-current-scene-update

Fix out of bounds access when updating current scene

YuriSizov committed:
Merge pull request #80037 from KoBeWi/we_are_reaching_levels_of_hacks_that_should_not_be_possible

Edit TileSet source on double click

YuriSizov committed:
Merge pull request #80066 from timothyqiu/no-pun-intended

Avoid punning `[param enabled]` in documentation

Faless committed:
[Web] Update npm packages

MewPurPur committed:
Add FileAccess and DirAccess icons

Spycemyster committed:
Added path to output of missing import texture metadata

timothyqiu committed:
Fix invalid use of markdown syntax in classref

Also:
* changed `[b]true[/b]` to `[code]true[/code]`
* use `[i]` for mathematical constant "e"
* use `[b]` for button text & menu item text
* improve markups about "tap1" and "tap2" in AudioEffectDelay

MewPurPur committed:
Add an icon to the Performancce object

Deledrius committed:
Fix typo in modulo error message.

akien-mga committed:
CI: Free disk space on Linux runners

Removing the Android toolchain saves 14 GiB, which gives us more room
for growth and to avoid running into out-of-space errors in the Linux
sanitizers + debug symbols builds.

Related to #79919, though the caches were just one part of the problem,
the real issue is that our Linux sanitizers builds take 12 GiB, and
adding godot-cpp on top with 2 GiB leaves only a few GiB left for the
cache itself.

Sauermann committed:
Actually store safe-rect in embedder

`sw` is a copy and not a reference.
Add unit-test for this case.

YuriSizov committed:
Merge pull request #69085 from marcinn/proposal-5748-loading-dds-at-runtime

Implement loading DDS textures at run-time

YuriSizov committed:
Merge pull request #77533 from EMBYRDEV/phys-mat-import-option

Add more physics options to the Scene importer

YuriSizov committed:
Merge pull request #79332 from AThousandShips/mp_crash

Prevent crash when accessing `Node` Multiplayer from thread

YuriSizov committed:
Merge pull request #79419 from KoBeWi/no_ID_for_you

Unexpose tile source ID

YuriSizov committed:
Merge pull request #79606 from clayjohn/ShaderRD-compilation-groups

Shader rd compilation groups

YuriSizov committed:
Merge pull request #79654 from timothyqiu/han

Fix script name of Hant and Hans

YuriSizov committed:
Merge pull request #79710 from AThousandShips/video_play

Fix incorrect virtual function in `VideoStream.set_paused`

YuriSizov committed:
Merge pull request #79776 from LRFLEW/mobilefog

Mobile: Uncomment code required for fog rendering on clear color

YuriSizov committed:
Merge pull request #79852 from YeldhamDev/that_was_getting_annoying

Fix usability issues with scene tabs

YuriSizov committed:
Merge pull request #79874 from clayjohn/PlaceholderTexture-rid

Lazily allocate RIDs for PlaceholderTextures to avoid allocating GPU resources unless used

YuriSizov committed:
Merge pull request #79911 from BastiaanOlij/fix_barrier_breakage

Add exceptions for breakage introduced in RD barriers

YuriSizov committed:
Merge pull request #79935 from dalexeev/gds-validate-node-path-annotation

GDScript: Add validation for `@export_node_path` annotation arguments

YuriSizov committed:
Merge pull request #79962 from viksl/update-tween-docs

Update C# example of `tween_method` with a parameter to the lambda method

YuriSizov committed:
Merge pull request #79990 from vnen/gdscript-assume-op-types

GDScript: Optimize operators by assuming the types

YuriSizov committed:
Merge pull request #80047 from AbeniMatteo/dev-colors

Mono: Avoid dictionary lookup for common colors

YuriSizov committed:
Merge pull request #80070 from clayjohn/GLES3-depth-prepass

Enable depth writes during shadow pass and depth pass. Disable during color pass

YuriSizov committed:
Merge pull request #80071 from darksylinc/matias-fixes2

Fix validation layer warnings

YuriSizov committed:
Merge pull request #80107 from Spycemyster/reimport-path-message

Add path to missing import texture metadata to error message

YuriSizov committed:
Merge pull request #80115 from akien-mga/ci-free-disk-space

CI: Free disk space on Linux runners

timothyqiu committed:
Fix "a number is required" error when printing RID

YuriSizov committed:
Merge pull request #80058 from dalexeev/editor-fix-pot-gen-escaping

Editor: Fix escaping issues with POT generator

Calinou committed:
Remove debugging print from shader cache

MewPurPur committed:
Add ShaderInclude icon

YuriSizov committed:
Merge pull request #67791 from Sauermann/fix-visual-mouse-notifications

Refactor mouse_entered and mouse_exited signals

YuriSizov committed:
Merge pull request #79258 from KoBeWi/drag_and_drop_anarchy

Assume root when dropping node to unassigned script

YuriSizov committed:
Merge pull request #79946 from Ymanawat/propagate-check-79942

Fix Tree check propagation not unchecking parent items

YuriSizov committed:
Merge pull request #79977 from Owl-A/bugfix

Add state sync after call to `_integrate_forces` in `_body_state_changed`

YuriSizov committed:
Merge pull request #79984 from KoBeWi/reverse_engineering_or_something_idk

Reverse condition for skipping directories

YuriSizov committed:
Merge pull request #80020 from dalexeev/gds-pot-gen-const-string

GDScript: Add constant string support for POT generator

YuriSizov committed:
Merge pull request #80109 from timothyqiu/not-markdown

Fix invalid use of markdown syntax in classref

YuriSizov committed:
Merge pull request #80122 from timothyqiu/rid-id

Fix "a number is required" error when printing RID

YuriSizov committed:
Merge pull request #80125 from Calinou/shader-cache-remove-debug-print

Remove debugging print from shader cache

YuriSizov committed:
CI: Extract godot-cpp testing into its own job

This ensures that the godot-cpp job has plenty of resources
to run its build and avoid being affected by the main build.

Additionally:
- Extract test tasks into dedicated actions.
- Upload artifacts as early as possible.
- Ensure that we check master cache before random cache.

aaronfranke committed:
Merge pull request #79907 from kdiduk/fix-advanced-scene-reimport-default-settings

Fix reimporting scene with default values selected

KoBeWi committed:
Ignore null "id" in tile source proxy

Sauermann committed:
Fix crash on Windows when closing `Window`

Send `WINDOW_EVENT_MOUSE_EXIT` when a window is destroyed.

Ensure, that on Windows, the event_callback is still valid during
destroying the `Window`.

akien-mga committed:
Merge pull request #80142 from Sauermann/fix-window-close-crash

Fix crash on Windows when closing `Window`

akien-mga committed:
Merge pull request #80091 from YuriSizov/ci-compartmentalization

CI: Extract godot-cpp testing into its own job

bruvzg committed:
[Windows] Do not force redraw window background on mouse pass-through region change.

lawnjelly committed:
Usage notes for DEV_ASSERT macro

Make it more clear that DEV_ASSERT is only to be used in specific cases.

RandomShaper committed:
Avoid crash on exiting due to late prints

akien-mga committed:
Merge pull request #69816 from Mickeon/doc-peeves-rect2m

Overhaul Rect2 & Rect2i Documentation

akien-mga committed:
Merge pull request #77280 from Rindbee/fix-unnecessary-break-in-Label

Fix unnecessary break when calculating the height of visible lines

akien-mga committed:
Merge pull request #78675 from AThousandShips/template_fix

Fix indentation in script templates

akien-mga committed:
Merge pull request #79249 from raulsntos/dotnet/dont-ignore-call-error

C#: Print error when MethodBind/Callable call fails

akien-mga committed:
Merge pull request #79267 from aaronfranke/gltf-copyright

Add copyright to GLTFState

akien-mga committed:
Merge pull request #79293 from ItsNL/add-accept-dialog-close-on-unfocused

Check `FLAG_POPUP` to close an AcceptDialog when parent is focused

akien-mga committed:
Merge pull request #79404 from raulsntos/dotnet/lines-open-in-external-editor

C#: Fix line in OpenInExternalEditor

akien-mga committed:
Merge pull request #79563 from danny88881/drag-and-drop-text-setting

Add drag'n'drop text option for `LineEdit` and `RichTextLabel`

akien-mga committed:
Merge pull request #79650 from garychia/filters_section

Avoid duplicating the "Filters" section

akien-mga committed:
Merge pull request #79705 from 398utubzyt/gui/snap-charfx-fix

GUI: Snap CharFX offset to nearest pixel when setting is enabled

akien-mga committed:
Merge pull request #79756 from sfreed141/root-node-casing-fix

In Create New Scene dialog derive the default root node name based on `editor/naming/node_name_casing`

akien-mga committed:
Merge pull request #79774 from aaronfranke/scene-import-root-type-name

Improve overriding the root type or root name in the scene importer

akien-mga committed:
Merge pull request #79775 from aaronfranke/gltf-node-gen-cosmetic

Cosmetic changes in GLTF node generation code

akien-mga committed:
Merge pull request #79801 from aaronfranke/gltf-scene-export-logic

Improve GLTF export logic for scene root nodes

akien-mga committed:
Merge pull request #80089 from AThousandShips/editor_settings_fix

Fix API hash related crash in `EditorSettings`

akien-mga committed:
Merge pull request #80103 from MewPurPur/access-icons

Add FileAccess and DirAccess icons

akien-mga committed:
Merge pull request #80113 from MewPurPur/performance-icon

Add an icon to the Performance object

akien-mga committed:
Merge pull request #80114 from Deledrius/modulo_typo

Fix typo in modulo error message.

akien-mga committed:
Merge pull request #80129 from MewPurPur/shader-include-icon

Add ShaderInclude class icon

akien-mga committed:
Merge pull request #80135 from KoBeWi/A_NumbER_Is_REQuIrEd

Ignore null "id" in tile source proxy

akien-mga committed:
Merge pull request #80153 from bruvzg/win_mouse_pass_redraw_fix

[Windows] Do not force redraw window background on mouse pass-through region change.

dalexeev committed:
GDScript: Fix regression with native signal not found

akien-mga committed:
Fix or workaround recent extension API compatibility issues

- Add compatibility methods for `RenderingDevice::shader_create_from_bytecode`
  and `CodeEdit::get_text_for_symbol_loopup`.
- Silence errors which now have compatibility methods.
- Acknowledge GraphEdit/GraphNode compat breakage, intended and WIP.

magian1127 committed:
Fix the error when clicking AnimationTree in the editor

YuriSizov committed:
Merge pull request #79321 from akien-mga/ci-allow-skipping-gha-on-forks

CI: Allow skipping our GHA workflows with `DISABLE_GODOT_CI` variable

YeldhamDev committed:
Expose `Window`'s `_get_contents_minimum_size()` to scripting

YuriSizov committed:
Merge pull request #79228 from smix8/fix_funnel_4.x

Fix pathfinding funnel adding unwanted point

YuriSizov committed:
Merge pull request #79501 from raulsntos/dotnet/warnings/NU5128

C#: Suppress NU5128 warning

YuriSizov committed:
Merge pull request #79588 from magian1127/4.0Fix78052

Fix the error when clicking AnimationTree in the editor

YuriSizov committed:
Merge pull request #79950 from bruvzg/win_app

[macOS Export] Disable unpacked .app bundle export on Windows.

YuriSizov committed:
Merge pull request #80092 from Faless/web/npm_update

[Web] Update npm packages

YuriSizov committed:
Merge pull request #80156 from lawnjelly/dev_assert_note

Usage notes for DEV_ASSERT macro

YuriSizov committed:
Merge pull request #80161 from RandomShaper/avoid_crash_on_late_prints

Avoid crash on exiting due to late prints

YuriSizov committed:
Merge pull request #80117 from Sauermann/fix-actually-store-safe-rect

Actually store safe-rect in embedder

Faolan-Rad committed:
XR Tracker reuses XR Pose

aaronfranke committed:
More cosmetic improvements in the GLTF code

raulsntos committed:
Show alert if .NET assemblies dir does not exist

akien-mga committed:
Merge pull request #71092 from KoBeWi/box_edit_level_up

Improve editing of box collision shapes

akien-mga committed:
Merge pull request #75198 from Bestest-Coder/master

Added --gpu-index to `forwardable_cli_arguments`

akien-mga committed:
Merge pull request #78108 from aaronfranke/variant-util-header

Make a header for VariantUtilityFunctions

akien-mga committed:
Merge pull request #78540 from anvilfolk/gdresloaderr

GDScript: Add error message when a GDScript resource fails to load.

akien-mga committed:
Merge pull request #79064 from MewPurPur/rename-button-pressed-THIS-IS-NOT-A-COMPAT-BREAK

Rename `button_pressed` default signal binding to avoid shadowing

akien-mga committed:
Merge pull request #79280 from raulsntos/dotnet/better-push-error

C#: Improve `GD.PushError` and `GD.PushWarning`

akien-mga committed:
Merge pull request #80165 from dalexeev/gds-fix-native-signal-not-found

GDScript: Fix regression with native signal not found

akien-mga committed:
Merge pull request #80178 from YeldhamDev/override_all_the_things

Expose `Window`'s `_get_contents_minimum_size()` to scripting

akien-mga committed:
Merge pull request #80198 from Faolan-Rad/XRos_Optimize

Optimized the XRTracker by reusing XRPose objects to minimize garbage collection overhead in C#

akien-mga committed:
Merge pull request #80205 from aaronfranke/gltf-cosmetic-2-electric-boogaloo

More cosmetic improvements in the GLTF code

akien-mga committed:
Merge pull request #80212 from raulsntos/dotnet/check_api_assemblies_dir_exists

Show alert if .NET assemblies dir does not exist

Mickeon committed:
Revert incorrect `Rect2.expand` description.

Also affects Rect2i of course.

akien-mga committed:
Merge pull request #80168 from akien-mga/extension-compat

Fix or workaround recent extension API compatibility issues

akien-mga committed:
Merge pull request #80217 from Mickeon/oops-rects

Revert incorrect `Rect2.expand` description.

YeldhamDev committed:
Improve `Window._get_contents_minimum_size()`'s code

akien-mga committed:
Merge pull request #80219 from YeldhamDev/override_all_the_things

Improve `Window._get_contents_minimum_size()`'s code

ettiSurreal committed:
Fix merge conflics, rename to rotate_toward

ettiSurreal commented:
Okay i have no idea what to do with this I'll just make a new PR.
